File: Django_Day/midterm APWP/midterm.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <algorithm>
#include <thread>

class Tank {
public:
    int damage;
    int hp;
    int armor;
    int price;
    int id;

    Tank(int d, int h, int a, int p, int i) : damage(d), hp(h), armor(a), price(p), id(i) {}
};

// Function to read tanks from the input file
std::vector<Tank> readTanksFromFile(const std::string &filename) {
    std::vector<Tank> tanks;
    std::ifstream file(filename);
    if (!file.is_open()) {
        std::cerr << "Error opening file." << std::endl;
        return tanks;
    }

    int numTanks;
    file >> numTanks;
    for (int i = 0; i < numTanks; ++i) {
        int damage, hp, armor, price;
        file >> damage >> hp >> armor >> price;
        tanks.emplace_back(damage, hp, armor, price, i + 1);
    }

    file.close();
    return tanks;
}

// Function to sort tanks by price and damage
void sortTanks(const std::vector<Tank> &tanks, std::vector<Tank> &sortedByPrice, std::vector<Tank> &sortedByDamage) {
    sortedByPrice = tanks;
    sortedByDamage = tanks;

    // Sort by price
    std::sort(sortedByPrice.begin(), sortedByPrice.end(), [](const Tank &a, const Tank &b) {
        return (a.price < b.price) || (a.price == b.price && a.damage > b.damage);
    });

    // Sort by damage
    std::sort(sortedByDamage.begin(), sortedByDamage.end(), [](const Tank &a, const Tank &b) {
        return (a.damage > b.damage) || (a.damage == b.damage && a.price < b.price);
    });
}

// Function to write results to the output file
void writeResultsToFile(const std::vector<Tank> &tanksPrice, const std::vector<Tank> &tanksDamage, const std::string &outputFilename) {
    std::ofstream file(outputFilename);
    if (!file.is_open()) {
        std::cerr << "Error opening output file." << std::endl;
        return;
    }

    file << "Sorted by Price:\n";
    for (const auto &tank : tanksPrice) {
        file << tank.id << '\n';
    }

    file << "\nSorted by Damage:\n";
    for (const auto &tank : tanksDamage) {
        file << tank.id << '\n';
    }

    file.close();
}

// Function to process tanks
void processTanks(const std::string &inputFilename, const std::string &outputFilename) {
    auto tanks = readTanksFromFile(inputFilename);
    std::vector<Tank> tanksSortedByPrice;
    std::vector<Tank> tanksSortedByDamage;
    
    sortTanks(tanks, tanksSortedByPrice, tanksSortedByDamage);
    writeResultsToFile(tanksSortedByPrice, tanksSortedByDamage, outputFilename);
}

// Main function
int main() {
    std::string inputFile = "input.txt";
    std::string outputFile = "output.txt";

    // Create a thread to process tanks
    std::thread tankThread(processTanks, inputFile, outputFile);
    tankThread.join();  // Wait for the thread to finish

    std::cout << "Tank processing completed. Results written to " << outputFile << "." << std::endl;
    return 0;
}

File: Django_Day/midterm APWP/midterm.py
import threading

class Tank:
    def __init__(self, damage, hp, armor, price, id):
        self.damage = damage
        self.hp = hp
        self.armor = armor
        self.price = price
        self.id = id

    def __repr__(self):
        return f'Tank(ID: {self.id}, Damage: {self.damage}, HP: {self.hp}, Armor: {self.armor}, Price: {self.price})'

def read_tanks_from_file(filename):
    tanks = []
    with open(filename, 'r') as file:
        num_tanks = int(file.readline().strip())
        for i in range(num_tanks):
            line = file.readline().strip().split()
            damage, hp, armor = map(int, line[:3])
            price = int(line[3])
            tanks.append(Tank(damage, hp, armor, price, i + 1))
    return tanks

def sort_tanks(tanks):
    # Sort by price
    tanks_sorted_by_price = sorted(tanks, key=lambda x: (x.price, -x.damage))
    # Sort by damage
    tanks_sorted_by_damage = sorted(tanks, key=lambda x: (-x.damage, x.price))
    return tanks_sorted_by_price, tanks_sorted_by_damage

def write_results_to_file(tanks_price, tanks_damage, output_filename):
    with open(output_filename, 'w') as file:
        file.write("Sorted by Price:\n")
        for tank in tanks_price:
            file.write(f'{tank.id}\n')
        
        file.write("\nSorted by Damage:\n")
        for tank in tanks_damage:
            file.write(f'{tank.id}\n')

def process_tanks(input_filename, output_filename):
    tanks = read_tanks_from_file(input_filename)
    tanks_sorted_by_price, tanks_sorted_by_damage = sort_tanks(tanks)
    write_results_to_file(tanks_sorted_by_price, tanks_sorted_by_damage, output_filename)

# Main thread
if __name__ == '__main__':
    input_file = 'input.txt'
    output_file = 'output.txt'

    # Create a thread to process tanks
    tank_thread = threading.Thread(target=process_tanks, args=(input_file, output_file))
    tank_thread.start()
    tank_thread.join()  # Wait for the thread to finish

    print(f'Tank processing completed. Results written to {output_file}.')